### PR TITLE
[couchbase] Add Integration Package with Bucket Data Stream

### PR DESCRIPTION
**Pull request by bhagyaraj-crest**
- Enhancement

## What does this PR do?

- Generated the skeleton of Couchbase integration package.
- Added 1 data stream ( bucket )
- Added data collection logic.
- Added the ingest pipelines.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
- Added system test cases.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's changelog.yml file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's manifest.yml file to point to the latest Elastic stack release (e.g. ^7.13.0).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/couchbase directory.
- Run the following command to run tests. 
> elastic-package test 

ss1
ss2

### CODE REVIEW

**Review by yug-elastic — APPROVED** [review]
LGTM!

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                  metric_type: gauge
+        - name: disk
+          type: group
+          fields:
+            - name: fetches
+              type: long
+              description: Number of disk fetches.
+              metric_type: gauge
+            - name: used
+              type: group
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:26]: Do we really need to make it group field? as it only contain one field, can we rather use used.bytes directly?

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+              fields:
+                - name: bytes
+                  type: long
+                  description: Amount of disk used (bytes).
+                  unit: byte
+                  metric_type: gauge
+        - name: item
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:33]: Do we really need to make it group field? as it only contain one field, can we rather use item.count directly?

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          type: group
+          fields:
+            - name: count
+              type: long
+              description: Number of items associated with the bucket.
+              metric_type: counter
+        - name: memory
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:40]: Do we really need to make it group field? as it only contain one field, can we rather use memory.used.bytes directly?

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          fields:
+            - name: quota
+              type: group
+              fields:
+                - name: bytes
+                  type: long
+                  description: Amount of RAM used by the bucket (bytes).
+                  unit: byte
+                  metric_type: gauge
+                - name: used
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:69]: Do we really need to make it group field? as it only contain one field, can we rather use used.pct directly?

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                  type: group
+                  fields:
+                    - name: pct
+                      type: float
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:73]: shouldn't this be scaled_float?

File: packages/couchbase/data_stream/bucket/fields/fields.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                - name: bytes
+                  type: long
+                  description: Amount of memory used by the bucket (bytes).
+                  unit: byte
+                  metric_type: gauge
+        - name: name
+          type: keyword
+          description: Name of the bucket.
+        - name: operations_per_sec
+          type: long
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/fields/fields.yml:55]: Please specify unit for long/float data types.

File: packages/couchbase/data_stream/bucket/sample_event.json [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            "type": "membase"
+        }
+    },
+    "data_stream": {
+        "dataset": "couchbase.bucket",
+        "namespace": "ep",
+        "type": "metrics"
+    },
+    "ecs": {
+        "version": "8.2.0"
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/sample_event.json:50]: The build ecs version is 8.3.0 and the sample event ecs version is 8.2.0.

File: packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+description: Pipeline for parsing Couchbase Bucket metrics.
+processors:
+  - set:
+      field: ecs.version
+      value: 8.3.0
+      ignore_failure: true
+      ignore_empty_value: true
+  - set:
+      field: event.type
+      value: [info]
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml:11]: nit: brackets not needed.

File: packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      ignore_failure: true
+      ignore_empty_value: true
+  - set:
+      field: event.kind
+      value: metric
+      ignore_failure: true
+      ignore_empty_value: true
+  - set:
+      field: event.category
+      value: [database]
```
Comment (ritalwar) [inline packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml:21]: nit: brackets not needed.

**Review by ritalwar — CHANGES_REQUESTED** [review]
Please re-check and confirm the couchbase logo that is being used. 
Also, I don't see any dashboards for this integration.

File: packages/couchbase/manifest.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+format_version: 1.0.0
+name: couchbase
+title: Couchbase
```
[the remark this answers is not in the document]
Comment (bhagyaraj-crest) [reply packages/couchbase/manifest.yml:3]: I believe most of the other integrations here are following the convention of mentioning the product name in the `title`. I just picked up a few common ones randomly.
References:
- https://github.com/elastic/integrations/blob/main/packages/1password/manifest.yml#L3
- https://github.com/elastic/integrations/blob/main/packages/apache/manifest.yml#L3
- https://github.com/elastic/integrations/blob/main/packages/google_workspace/manifest.yml#L2

Please let me know if you still think updating it `Couchbase Metrics` sounds more intuitive to you.

File: packages/couchbase/manifest.yml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    size: 216x216
+    type: image/svg+xml
+policy_templates:
+  - name: couchbase
+    title: Couchbase metrics
+    description: Collect Couchbase metrics.
+    inputs:
+      - type: http/metrics
+        title: Collect Couchbase metrics using http module
+        description: Collect Couchbase metrics using http module.
```
[the remark this answers is not in the document]
Comment (bhagyaraj-crest) [reply packages/couchbase/manifest.yml:25]: Agreed. However, the reason for keeping it generic is keeping in mind it's future extensibility of the data streams.
One more thing to add here, we already have the data stream specific title nested inside as the toggle button for more details. Hence, mentioning the details in the main title would make it redundant in my opinion. Please refer the screenshot.
image
These were just my two cents. Do you think we should still update the `title` as well as `description` to: `Collect bucket/cluster metrics from Couchbase instances`?

File: packages/couchbase/data_stream/bucket/manifest.yml [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+          #   MDEtQkFTRS05MDkzMzJjMC1iNmFiLTQ2OTMtYWI5NC01Mjc1ZDI1MmFmNDiCCWxv
+          #   Y2FsaG9zdDARBgNVHQ4ECgQITzqhA5sO8O4wDQYJKoZIhvcNAQELBQADggEBAKR0
+          #   gY/BM69S6BDyWp5dxcpmZ9FS783FBbdUXjVtTkQno+oYURDrhCdsfTLYtqUlP4J4
+          #   CHoskP+MwJjRIoKhPVQMv14Q4VC2J9coYXnePhFjE+6MaZbTjq9WaekGrpKkMaQA
+          #   iQt5b67jo7y63CZKIo9yBvs7sxODQzDn3wZwyux2vPegXSaTHR/rop/s/mPk3YTS
+          #   hQprs/IVtPoWU4/TsDN3gIlrAYGbcs29CAt5q9MfzkMmKsuDkTZD0ry42VjxjAmk
+          #   xw23l/k8RoD1wRWaDVbgpjwSzt+kl+vJE/ip2w3h69eEZ9wbo6scRO5lCO2JM4Pr
+          #   7RhLQyWn2u00L7/9Omw=
+          #   -----END CERTIFICATE-----
+        description: i.e. certificate_authorities, supported_protocols, verification_mode etc.
```
[the remark this answers is not in the document]
Comment (bhagyaraj-crest) [reply packages/couchbase/data_stream/bucket/manifest.yml:48]: We had a similar discussion over other GitHub PR with @muthu-mps and we agreed upon keeping it consistent with the other integrations.
Please refer: https://github.com/elastic/integrations/pull/3591#discussion_r916613240

File: packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml (hunk 9 above)
Comment (bhagyaraj-crest) [reply packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml:21]: Please refer to the note mentioned in the ECS field reference for the field.
https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-category

> Note: this field should contain an array of values.

File: packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml (hunk 8 above)
Comment (bhagyaraj-crest) [reply packages/couchbase/data_stream/bucket/elasticsearch/ingest_pipeline/default.yml:11]: Please refer to the note mentioned in the ECS field reference for the field.
https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-type

> Note: this field should contain an array of values.

**Comment by bhagyaraj-crest** [comment]
> Please re-check and confirm the couchbase logo that is being used. 

Other option is to use the following logo:
https://worldvectorlogo.com/logo/couchbase
If you have any other link to a logo, kindly share and we can update it.


> Also, I don't see any dashboards for this integration.

Kindly refer the discussion here: https://github.com/elastic/integrations/pull/3706#issuecomment-1188991096

File: packages/couchbase/data_stream/bucket/fields/fields.yml (hunk 6 above)
Comment (ritalwar) [reply packages/couchbase/data_stream/bucket/fields/fields.yml:37]: No, let's not.

**Comment by ritalwar** [comment]
> https://worldvectorlogo.com/logo/couchbase

Yes, let's go with this one.

**Comment by bhagyaraj-crest** [comment]
> > https://worldvectorlogo.com/logo/couchbase
> 
> Yes, let's go with this one.

Sure, thanks for the confirmation. Updated the PR with the changes. Kindly perform a final round of review and let us know if you have any further concerns.

**Pull request merged**